osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsGasCommandExecutor.java

```diff
@@ -291,318 +131,199 @@ public PeriodicMeterReadGasResponseDto execute(final DlmsConnectionManager conn,
+                periodicMeterReadsQuery.getChannel().getChannelNumber());
+        final Date captureTime = this.readCaptureTime(bufferedObjects, selectedObjects);
 
-        final CosemDateTimeDto cosemDateTime = this.dlmsHelper
-                .readDateTime(bufferedObjects.get(BUFFER_INDEX_CLOCK), "Clock from " + queryPeriodType + " buffer gas");
-        final DateTime bufferedDateTime = cosemDateTime == null ? null : cosemDateTime.asDateTime();
+        LOGGER.info("Converting bufferObject with value: {} ", bufferedObjects);
+        LOGGER.info("Resulting values: LogTime: {}, status: {}, gasValue {}, scalerUnit: {}, captureTime {} ", logTime,
+                status, gasValue, scalerUnit, captureTime);
 
-        this.dlmsHelper.validateBufferedDateTime(bufferedDateTime, cosemDateTime, queryBeginDateTime, queryEndDateTime);
+        return new PeriodicMeterReadsGasResponseItemDto(logTime,
+                this.dlmsHelper.getScaledMeterValue(gasValue, scalerUnit, GAS_VALUE), captureTime, status);
+    }
+
+    private Integer getIndex(final List<DlmsCaptureObject> selectedObjects, final DlmsObjectType type,
+            final Integer attributeId) {
+        int index = 0;
 
-        LOGGER.debug("Processing profile ({}) objects captured at: {}", queryPeriodType, cosemDateTime);
+        for (final DlmsCaptureObject object : selectedObjects) {
+            if (object.getRelatedObject().getType().equals(type) && (attributeId == null
+                    || object.getAttributeId() == attributeId)) {
+                return index;
+            }
+            index++;
+        }
 
-        return this.getNextPeriodicMeterReadsBasedOnPeriodType(periodicMeterReadsQuery, bufferedObjects,
-                isSelectiveAccessSupported, getResultList, bufferedDateTime);
+        return null;
     }
 
-    private PeriodicMeterReadsGasResponseItemDto getNextPeriodicMeterReadsBasedOnPeriodType(
-            final PeriodicMeterReadsRequestDto periodicMeterReadsQuery, final List<DataObject> bufferedObjects,
-            final boolean isSelectiveAccessSupported, final List<GetResult> getResultList,
-            final DateTime bufferedDateTime) throws ProtocolAdapterException, AssertionError {
+    private Date readClock(final PeriodicMeterReadsRequestDto periodicMeterReadsQuery,
+            final List<DataObject> bufferedObjects, final List<DlmsCaptureObject> selectedObjects)
+            throws ProtocolAdapterException, BufferedDateTimeValidationException {
 
         final PeriodTypeDto queryPeriodType = periodicMeterReadsQuery.getPeriodType();
-        final ChannelDto queryChannel = periodicMeterReadsQuery.getChannel();
+        final DateTime queryBeginDateTime = new DateTime(periodicMeterReadsQuery.getBeginDate());
+        final DateTime queryEndDateTime = new DateTime(periodicMeterReadsQuery.getEndDate());
 
-        switch (queryPeriodType) {
```

Comment:
Suggest to move this switch into either of the enums it is mapping, where it can be implemented as a lookup on the enum values which get a single arg constructor with the mapped value.

Reply:
Aangepast